erts/emulator/beam/erl_map.c

```diff
@@ -368,10 +369,14 @@ static Eterm flatmap_from_validated_list(Process *p, Eterm list, Eterm fill_valu
     Sint  idx = 0;
 
 
-    hp    = HAlloc(p, 3 + 1 + (2 * size));
+    hp    = HAlloc(p, 3 + (size == 0 ? 0 : 1) + (2 * size));
```

Comment:
If I understand correctly this is to safely handle empty maps. If that's the case, would it make sense to also apply the same treatment to empty maps and have just one empty map literal that could be returned?

Reply:
Yes, we would save some memory and perhaps utilize the processor caches better by doing that. The drawback would be that, in some places, we would need a little bit more code when creating maps in BIF's to check if the size is 0 and then return the empty map (e.g., the [maps:remove/2 BIF](https://github.com/erlang/otp/blob/7079bda1d36707788149197ec75238cbc8c7b6b2/erts/emulator/beam/erl_map.c#L1834)). 

The empty tuple optimization was primary motivated by the fact that it (together with padding literal areas with 1 word) enables us to ensure that the word after the arity word of boxed terms always is allocated memory which in turn enables some optimizations in the ARM JIT and for invocations of funs.

However, an empty map takes up more space than an empty tuple and empty maps are probably much more common than empty tuples so I think it can be worth it just because it gives better memory properties. An alternative (suggested by @jhogberg ) to creating a constant empty map is to introduce a new term type for the empty map similar to how we do with the empty list.